epix/io.py

```diff
@@ -196,6 +197,139 @@ def _get_ttree(directory, file_name):
     return ttree, n_simulated_events
 
 
+def csv_loader(directory,
+              file_name,
+              arg_debug=False,
+              outer_cylinder=None,
+              kwargs={},
+              cut_by_eventid=False,
+              ):
+    """
+    Function which loads handcrafted instructions from a csv file unsing pandas.
+
+    Args:
+        directory (str): Directory in which the data is stored.
+        file_name (str): File name
+        arg_debug (bool): If true, print out loading information.
+        kwargs (dict): Keyword arguments similar to kwargs_uproot_arrays in loader function
+        cut_by_eventid (bool): If true event start/stop are applied to
+            eventids, instead of rows.
+        
+    Returns:
+        awkward1.records: Interactions (eventids, parameters, types).
+        integer: Number of events simulated.
+    """
+    instr_df =  pd.read_csv(os.path.join(directory, file_name))
+    instr_df["r"] = np.sqrt(instr_df["x"]**2 + instr_df["y"]**2) 
+    
+    n_simulated_events = len(np.unique(instr_df.eventid))
+    
+    if arg_debug:
+        print(f'Total entries in input file = {len(instr_df)}')
+        cutby_string='output file entry'
+        if cut_by_eventid:
+            cutby_string='g4 eventid'
+
+        if kwargs['entry_start'] is not None:
+            print(f'Starting to read from {cutby_string} {kwargs["entry_start"]}')
+        if kwargs['entry_stop'] is not None:
+            print(f'Ending read in at {cutby_string} {kwargs["entry_stop"]}')
+
+    # If user specified entry start/stop we have to update number of
+    # events for source rate computation:
+    if kwargs['entry_start'] is not None:
+        start = kwargs['entry_start']
+    else:
+        start = 0
+
+    if kwargs['entry_stop'] is not None:
+        stop = kwargs['entry_stop']
+    else:
+        stop = n_simulated_events
+    n_simulated_events = stop - start
+
+    if cut_by_eventid:
+        # Start/stop refers to eventid so drop start drop from kwargs
+        # dict if specified, otherwise we cut again on rows.
+        kwargs.pop('entry_start', None)
+        kwargs.pop('entry_stop', None)
+
+
+    if outer_cylinder:
+        cut_string = (f'(r < {outer_cylinder["max_r"]})'
+                      f' & ((z >= {outer_cylinder["min_z"] * 10}) & (z < {outer_cylinder["max_z"] * 10}))')
+        instr_df = instr_df.query(cut_string)
```

Comment:
These are the type of things which may diverge at some point if not paid attention to. Can we find a general solution for both file types?

Reply:
I moved the cut_string in this commit: https://github.com/XENONnT/epix/pull/42/commits/be082ba449522d8217c6fcf1f6bde79f4638aa15. This change now requires that the input csv file contains the positions in mm as in the root file.